src/sql/workbench/api/node/mainThreadNotebookDocumentsAndEditors.ts

```diff
@@ -0,0 +1,367 @@
+import { IExtHostContext } from 'vs/workbench/api/node/extHost.protocol';
+import { IEditorService } from 'vs/workbench/services/editor/common/editorService';
+import { IEditorGroupsService } from 'vs/workbench/services/group/common/editorGroupsService';
+import { IInstantiationService } from 'vs/platform/instantiation/common/instantiation';
+import { ITextEditorOptions } from 'vs/platform/editor/common/editor';
+import { viewColumnToEditorGroup } from 'vs/workbench/api/shared/editor';
+
+import {
+	SqlMainContext, MainThreadNotebookDocumentsAndEditorsShape, SqlExtHostContext, ExtHostNotebookDocumentsAndEditorsShape,
+	INotebookDocumentsAndEditorsDelta, INotebookEditorAddData, INotebookShowOptions, INotebookModelAddedData
+} from 'sql/workbench/api/node/sqlExtHost.protocol';
+import { NotebookInputModel, NotebookInput } from 'sql/parts/notebook/notebookInput';
+import { INotebookService, INotebookEditor } from 'sql/services/notebook/notebookService';
+import { TPromise } from 'vs/base/common/winjs.base';
+
+class MainThreadNotebookEditor extends Disposable {
+
+	constructor(public readonly editor: INotebookEditor) {
+		super();
+	}
+
+	public get uri(): URI {
+		return this.editor.notebookParams.notebookUri;
+	}
+
+	public get id(): string {
+		return this.editor.id;
+	}
+
+	public get isDirty(): boolean {
+		return this.editor.isDirty();
+	}
+
+	public get providerId(): string {
+		return this.editor.notebookParams.providerId;
+	}
+
+	public save(): Thenable<boolean> {
+		return this.editor.save();
+	}
+
+	public matches(input: NotebookInput): boolean {
+		if (!input) {
+			return false;
+		}
+		return input === this.editor.notebookParams.input;
+	}
+}
+
+function wait(timeMs: number): Promise<void> {
+	return new Promise(resolve => setTimeout(resolve, timeMs));
+}
+
+
+namespace mapset {
+
+	export function setValues<T>(set: Set<T>): T[] {
+		// return Array.from(set);
+		let ret: T[] = [];
+		set.forEach(v => ret.push(v));
+		return ret;
+	}
+
+	export function mapValues<T>(map: Map<any, T>): T[] {
+		// return Array.from(map.values());
+		let ret: T[] = [];
+		map.forEach(v => ret.push(v));
+		return ret;
+	}
+}
+
+namespace delta {
+
+	export function ofSets<T>(before: Set<T>, after: Set<T>): { removed: T[], added: T[] } {
+		const removed: T[] = [];
+		const added: T[] = [];
+		before.forEach(element => {
+			if (!after.has(element)) {
+				removed.push(element);
+			}
+		});
+		after.forEach(element => {
+			if (!before.has(element)) {
+				added.push(element);
+			}
+		});
+		return { removed, added };
+	}
+
+	export function ofMaps<K, V>(before: Map<K, V>, after: Map<K, V>): { removed: V[], added: V[] } {
+		const removed: V[] = [];
+		const added: V[] = [];
+		before.forEach((value, index) => {
+			if (!after.has(index)) {
+				removed.push(value);
+			}
+		});
+		after.forEach((value, index) => {
+			if (!before.has(index)) {
+				added.push(value);
+			}
+		});
+		return { removed, added };
+	}
+}
+
+class NotebookEditorStateDelta {
```

Comment:
All the state code is based on MainThreadDocumentsAndEditors, but is technically net-new since I've had to tweak things to work. In particular, we don't have separate document tracking.